rpc/Transceiver.java

```diff
@@ -82,69 +78,36 @@ public static Transceiver create(SessionServiceGrpc.SessionServiceStub stub) {
      * Send a request and return immediately.
          * This method is non-blocking - it returns immediately.
      */
-    private void send(Transaction.Req request) {
-        try (ThreadTrace trace = traceOnThread("request")) {
-            if (responseListener.terminated.get()) {
-                throw GraknClientException.connectionClosed();
-            }
-            LOG.trace("send:{}", request);
-            requestSender.onNext(request);
+    private void send(Transaction.Req request, ResponseCollector collector) {
+        if (responseListener.terminated.get()) {
+            throw GraknClientException.connectionClosed();
         }
-    }
+        LOG.trace("send:{}", request);
 
-    /**
-     * Block until a response is returned.
-     */
-    private Response receive() throws InterruptedException {
-        try (ThreadTrace trace = traceOnThread("receive")) {
-            Response response = responseListener.poll();
-            LOG.trace("receive:{}", response);
-            if (response.type() != Response.Type.OK) {
-                close();
-            }
-            return response;
+        // We must add the response collectors in exact the same order we send the requests
+        synchronized (this) {
+            responseListener.addCollector(collector); // Must add collector first to be watertight
+            requestSender.onNext(request);
```

Comment:
how does synchronising these two lines guarantee that the order, @adammitchelldev ? Don't you need to make the method to be `synchronised` in order to make the order guaranteed?

Reply:
and do we need to add a test for this?

Reply:
The `synchronized (this)` acquires an implicit monitor lock for this block, so only one thread can ever call this block of code on this `Transceiver` at a time. This is the point at which we send the actual request, so by synchronizing this with recording the collector (corresponding to this request), we guarantee that we'll be processing the results in the same order we send the requests.

Nothing else is conflicting, so nothing else needs to be synchronized.

Reply:
Worth bearing in mind that
```
public synchronized void doSync() {
    // code
}
```
Is the same as:
```
public void doSync() {
    synchronized (this) {
        // code
    }
}
```
`synchronized (this)` obtains an object-wide lock.

Reply:
Yes I know they're the same, but judging from your comment, you said that you wanted to maintain order or based on who called `send()` first? If so, then adding the `synchronized` inside the function won't guarantee it being executed in the order of who called the `send()` method, right?

Reply:
Ah, the order of who calls the outer send method isn't important, only that whichever one gets given to GRPC first also has its collector added first. If two different threads were to make requests at the same time, there would be no guarantee which order they'd actually get sent in either way.

Reply:
Ah, right. Okay merging this.

Reply:
By send I mean send the GRPC request (`requestSender.onNext(request);`)